### ISSUE TITLE
Integration error depending on IA amplitude

### ISSUE THREAD

**Issue by joezuntz** [LSSTDESC/CCL#302]
I'm seeing an integration error when calculating C_ell for the DES SV n(z) that depends on the value of the IA amplitude, which seems a bit odd.

I wanted to double check if I'm doing something wrong with ranges or input values or something else like that.

I've attached an example which uses the current CCL master branch.  I had to call it issue.py.txt so that github would let me attach it.  It loads the attached number densities.

If you use A=1.0 then it works, but if you use A=3.0 then it fails with  
```
RuntimeError: Error CCL_ERROR_INTEG: ccl_cls.c: ccl_angular_cl(): error integrating over k
```
Can anyone advise?

[bin_1.txt](https://github.com/LSSTDESC/CCL/files/1556534/bin_1.txt)
[bin_3.txt](https://github.com/LSSTDESC/CCL/files/1556535/bin_3.txt)
[z.txt](https://github.com/LSSTDESC/CCL/files/1556536/z.txt)
[issue.py.txt](https://github.com/LSSTDESC/CCL/files/1556538/issue.py.txt)

**Comment by joezuntz** [comment]
Also i was giving it a very wide range in ell as I thought that this would be needed for the conversion to xi(theta).  If that's not the case then let me know.

**Comment by elisachisari** [comment]
@joezuntz The "intrinsic alignment bias" as defined in CCL is not normalized following usual conventions in IA literature. See issue #280. I thought I'd point this out. I think you are passing a huge IA amplitude if you set IAbias = 3 and this is somehow causing integration problems (which we should investigate).

**Comment by joezuntz** [comment]
@elisachisari thanks!  Can you advise how I should change the code attached above to make it sensibly normalized?  I included the C_1 factor already.

**Comment by jablazek** [comment]
@elisachisari : assuming the IAbias is linear in A, then it is strange if A=3 causes problems when A=1 does not. But either way, yes we should address #280.

**Comment by jablazek** [comment]
From issue.py above:
> bias_ia = -A * C1_rhocrit * Omega_m / D

This looks "standard." @elisachisari : is something happening downstream that rescales this bias, or is it bias_ia or bias_ia^2 that multiplies P_delta in the NLA model?

**Comment by damonge** [comment]
I have to look into this for a different project, so let me get back to you today

**Comment by damonge** [comment]
@joezuntz : alright, the way this is implemented in CCL is as follows (I'm quoting from https://arxiv.org/pdf/1506.02671.pdf, eqs. 4.14, 5.2):

1. In the large-ell limit, the IA kernel is basically proportional to W_IA =  p_red*b_I, where p_red(z) is the redshift distribution of the galaxy population that aligns, and b_I(z) is some alignment bias function.
2. Writing p_red(z) = p_tot(z)*f_red(z), where f_red is the "red" fraction, this is just W_IA = p_tot(z)*f_red(z)*b_I(z). Currently users need to pass both b_I and f_red to CCL (as was initially suggested by @elisachisari and @jablazek).
3. According to Eq. 5.2 of the paper above, b_I can be related to the more common A_I as: b_I = -A_I*Omega_M*C_1*rho_crit/D(z)

I'm gonna copy this into #280 . If you confirm that this is really the parametrization that you want I could easily implement the modification in the code. Not sure whether this will solve the integration error, but it doesn't hurt to try.

**Comment by elisachisari** [comment]
Ok, so you are accounting for the amplitude normalization then. There has to be some numerical convergence problem with the Cl's for large A. I would try varying CCL spline parameters to see if this has an impact.
@damonge @jablazek: I would be in favor of incorporating this IA normalization directly in CCL, as it is standard in cosmic shear papers.

**Comment by elisachisari** [comment]
@jablazek IA is linear in A but the cosmic shear prediction @joezuntz is doing is not. If IA and shear terms are summed before integrating in k, then it could be that this triggers numerical integration problems. @damonge would know the exact implementation.

**Comment by jablazek** [comment]
@elisachisari : yes, that sounds plausible. If the k->ell integral is all done in one step by first summing the relevant P(k) with their amplitudes and kernels, then this would make sense.

 A=3 is indeed large for lensing sources, but not large for LRGs (which someone using CCL may want to consider).

**Comment by elisachisari** [comment]
We might consider changing the implementation: for example, integrating the C_ell contribution of IA and lensing (auto and cross) separately, and then adding C_ells. I think this could be more stable.

**Comment by joezuntz** [comment]
That way of doing it is also useful if you want to be able to see the different contributions from the different components.  It is generally slower though as you have to do all the integrals again.

**Comment by EiffL** [comment]
@damonge @elisachisari @jablazek Ok, I've looked into the issue, the culprit is here:
https://github.com/LSSTDESC/CCL/blob/b778440fba24252fdb9818480f81057368dd6800/src/ccl_cls.c#L742
The 4th argument in this gsl integration function is the `abserr` requested and is currently set to 0. I didn't find in the GSL documentation what happens when you set `abserr` to zero but I'll bet that it tries to compute the integral down to numerical accuracy in absolute error, i.e. 1e-14, I can work sometimes but not all the times.

When I run the code as-is, in the case that @joezuntz was interested in, I get an `GSL_EROUND: cannot reach tolerance because of roundoff error, or roundoff error was detected in the extrapolation table.` error code, but only for one particular `ell` in the array, suggesting just some numerical accuracy issue. 

When relaxing `abserr` to an arbitrary `1e-9` everything works, I can even set `A_I` to 100 without any issues, and the relative error remains around `1e-10` even in this case.

So, two questions:
  - Is there a good reason that I don't know about for setting `abserr` to zero in `gsl_integration_qag` ?
  - Is there a convention in CCL for the numerical errors permitted in this kind of integrations ?

Also, @ssamuroff is looking into separating the contributions of IA in the C_ell as also suggested in this thread.

**Comment by NiallMac** [comment]
@EiffL In case at all useful, we had some sort of related confusion with abserr and epsrel in the Limber integration routines in cosmosis. I settled on setting abserr=0 by default since in this case (according to the gsl documentation) the gsl integration routines should just use epsrel, and given the wide range in absolute values different types of angular spectra can have, it wasn't obvious to me what a sensible default non-zero value of abserr was. It is possible though that this choice will run into problems for v. small integrals where epsrel may approach machine epsilon...

**Comment by EiffL** [comment]
@NiallMac you were right, I didn't know about it but after a little more digging it turns out that indeed setting `epsabs=0` forces the integration function to consider only the relative error, so that's the right thing to do.

I have experimented a little bit more and my understanding is that the integration fails just in some edge cases because of round-off errors that occur because the `gsl_integration_qag` function sums a lot of positive and negative values with roughly the same amplitude. I have added in PR #353 a test for round-off errors and if the integral fails, then another more robust function `gsl_integration_cquad` is used as backup. In my tests, cquad was able to compute the integral in all the cases where the normal qag failed, with whatever IA amplitude.
